pkg/parse/source.go

```diff
@@ -95,10 +95,33 @@ func (o *files) readConfigFiles(state *sourceState) status.Error {
 	if err != nil {
 		return status.PathWrapError(errors.Wrap(err, "listing files in the configs directory"), syncDir.OSPath())
 	}
+
+	commitChanged, newCommit, err := o.checkCommitChanged(p, state)
+	if err != nil {
+		return status.TransientError(err)
+	} else if commitChanged && newCommit != "" {
+		return status.TransientError(fmt.Errorf("source commit changed while listing files, was %s, now %s. It will be retried in the next sync", state.commit, newCommit))
+	}
+
 	state.files = fileList
 	return nil
 }
 
+// checkCommitChanged returns true and the updated commit if the computed commit
+// is different from the given commit in sourceState, or returns error if found
+// during computation
+func (o *files) checkCommitChanged(p Parser, state *sourceState) (bool, string, error) {
+	gs := sourceStatus{}
+	gs.commit, _, gs.errs = hydrate.SourceCommitAndDir(p.options().SourceType, p.options().SourceDir, p.options().SyncDir, p.options().reconcilerName)
```

Comment:
Seems like this could be significantly simplified. The commit is just this:
```
gitDir, err := sourceRevDir.EvalSymlinks()
	if err != nil {
		return "", "", status.SourceError.Sprintf("unable to evaluate the source link %s: %v", sourceRevDir, err).Build()
	}

	commit := filepath.Base(gitDir.OSPath())
```

Maybe make a smaller `hydrate.SourceCommit` function.

Reply:
Or do we care what the error file says and want that to interrupt this?

Reply:
I like the idea. We don't need to recheck the existence of the mounted directory and the error file. All we need is to re-evaluate the symlink.

Reply:
Anyway, my motivation here is to not pass the full parser into checkCommitChanged. You only needs the "options" to find the commit, right?

Reply:
> Or do we care what the error file says and want that to interrupt this?

I don't think we care. It might be the error from the previous commit. If it is the error for the current commit, it will be returned in the next retry.

Reply:
Now passing the sourceRevDir only